client/go/dogma/dogma.go

```diff
@@ -0,0 +1,375 @@
+	c.repository = (*repositoryService)(service)
+	c.content = (*contentService)(service)
+	return c, nil
+}
+
+func normalizeURL(baseURL string) (*url.URL, error) {
+	if len(baseURL) == 0 {
+		return url.Parse(defaultBaseURL)
+	}
+
+	if !strings.HasPrefix(baseURL, "http") {
+		// Prepend the defaultScheme when there is no specified scheme so parse the url properly
+		// in case of "hostname:port".
+		baseURL = defaultScheme + "://" + baseURL
+	}
+
+	parsedURL, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	if len(parsedURL.Scheme) == 0 {
+		baseURL = defaultScheme + "://" + baseURL
+	}
+	port := parsedURL.Port()
+	if len(port) == 0 {
+		baseURL += ":" + strconv.Itoa(DefaultPort)
+	}
+	if !strings.HasSuffix(baseURL, "/") {
+		baseURL += "/"
+	}
+	return url.Parse(baseURL)
+}
+
+// SecurityEnabled returns whether the security of the server is enabled or not.
+func (c *Client) SecurityEnabled() (bool, error) {
+	u, err := c.baseURL.Parse(pathSecurityEnabled)
+	if err != nil {
+		return false, err
+	}
+
+	req := &http.Request{Method: http.MethodGet, URL: u}
+	res, err := c.client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("authenticaion failed (status: %s)", res.Status)
+	}
+	b, _ := ioutil.ReadAll(res.Body)
+	if string(b) == "true" {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (c *Client) newRequest(method, urlStr string, body interface{}) (*http.Request, error) {
+	u, err := c.baseURL.Parse(urlStr)
+	if err != nil {
+		return nil, err
+	}
+
+	var buf io.ReadWriter
+	if body != nil {
+		if str, ok := body.(string); ok {
+			buf = bytes.NewBufferString(str)
+		} else {
+			buf = new(bytes.Buffer)
+			enc := json.NewEncoder(buf)
+			enc.SetEscapeHTML(true)
+			err := enc.Encode(body)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	req, err := http.NewRequest(method, u.String(), buf)
+	if err != nil {
+		return nil, err
+	}
+
+	if auth := req.Header.Get("Authorization"); len(auth) == 0 {
+		req.Header.Set("Authorization", "Bearer anonymous")
+	}
+
+	if body != nil {
+		if method == "PATCH" {
+			req.Header.Set("Content-Type", "application/json-patch+json")
+		} else {
+			req.Header.Set("Content-Type", "application/json")
+		}
+	}
+	return req, nil
+}
+
+func (c *Client) do(ctx context.Context, req *http.Request, resContent interface{}) (*http.Response, error) {
+	req = req.WithContext(ctx)
+
+	res, err := c.client.Do(req)
+	if err != nil {
+		fmt.Println(err)
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+	}
+	defer res.Body.Close()
+
+	if resContent != nil {
+		err = json.NewDecoder(res.Body).Decode(resContent)
+		if err == io.EOF { // empty response body
+			err = nil
+		}
+	}
+
+	return res, err
+}
+
+// CreateProject creates a project.
+func (c *Client) CreateProject(ctx context.Context, name string) (*Project, *http.Response, error) {
+	return c.project.create(ctx, name)
```

Comment:
Not sure what is better, but compare between below 2 interfaces, I prefer former one. HDYT?

```
// expose service object
dogma.Project.Create(...)
```

vs

```
// current approach
dogma.CreateProject(...)
```

Reply:
> same method name as in Java client 
this sound reasonable, so let's keep compability for emthod naming rule